### ISSUE TITLE
The canonical definition of AwaitTaskCorrect is incorrect

### ISSUE THREAD

**Issue by deviousasti** [jet/equinox#336]
Consider the following case of 
- short-lived `Async`: **A**
- long running `Task` awaited by `AwaitTaskCorrect `: **B** 
- long running `Async`: **C**

Async **A** will fault when it counts do to zero from 5.

```fsharp
let rec loop i : Async<unit> = async {        
        do! Async.Sleep 500        
        do printfn $"I'm alive! But not for long: {i}"
        if i = 0 then invalidOp "Is zero even a number?"
        return! loop (i - 1)
}
let A = loop 5

let B = Task.Delay Timeout.Infinite |> Async.AwaitTaskCorrect

let C = Async.Sleep Timeout.Infinite

[<EntryPoint>]
let main _argv =     
    Async.Parallel [ A; B ]
    |> Async.Ignore
    |> Async.RunSynchronously
    0
```

### Expectation
Since **A** will fault when i = 0, we expect the error bubble up un-handled to Async.Parallel and subsequently to main.
The program should crash in 3 seconds.

### Observed
**A** faulted - the eventloop terminates, but the program does not. 
The program is silently left in an inconsistent state.

### Works as advertised
Barring implementation details, **C** is functionally equivalent to **B**.
If we substitute **C** for **B**:
```
Async.Parallel [ A; C ]
```
the program will terminate as expected.

**Comment by bartelink** [comment]
`Async.Parallel` and `AwaitTaskCorrect` will never leave an orphan task running.

I can't recall which `dotnet/fsharp` issue covers this, but the problem here is not `AwaitTaskCorrect` - it's the fact that the `Delay` you are invoking is working as spec'd -- delaying forever. `Async.Sleep`, on the other hand, internally honors cancellation (because it can, via `Async.CancellationToken`).

Thus I'd claim this is an invalid case - if you replace B with
```
let B = async {
    let! ct = Async.CancellationToken
    return! Task.Delay(Timeout.Infinite, cancellationToken = ct) |> Async.AwaitTaskCorrect
}
```

I'm pretty sure `Async.AwaitTask` shares the same semantics.

**Comment by deviousasti** [comment]
Yes, `Async.AwaitTask` shares the same semantics. Which is why I'd expect `AwaitTaskCorrect` to improve on the semantics.
It's same story for something based on anything based on TaskCompletionSource.

**Comment by bartelink** [comment]
Leaking unawaited `Task`s on every cancellation is an acquired taste; I'd be on the side of preferring the determinism at all costs.
Things like `Task.AwaitAll` don't do such things
Yes, there is an argument that this can be painful in practice - e.g. like the .NET 4.0 crashing the process if a Task is unobserved, but I think the issues in `dotnet/fsharp` have traversed this territory far better than we'll ever do here (some are still open - it's definitely not an open and shut thing)

Bottom line in the context of the `github/jet` repos is that they're all designed to play well with these semantics, e.g. https://github.com/jet/propulsion/blob/master/src/Propulsion/Internal.fs#L68

**Comment by deviousasti** [comment]
I will fix this footgun in user-land. (-__-)
